File: Frontend/src/types/contracts/preparation.ts
import { z } from "zod";

import { difficultySchema } from "@/types/contracts/common";

export const preparationTaskSchema = z.object({
  id: z.string(),
  day: z.number(),
  dateLabel: z.string(),
  phase: z.string(),
  category: z.string(),
  title: z.string(),
  description: z.string(),
  estimatedMinutes: z.number(),
  status: z.enum(["pending", "in_progress", "completed"]),
  priority: z.enum(["high", "normal"]),
});

export const questionSchema = z.object({
  id: z.string(),
  text: z.string(),
  category: z.string(),
  topic: z.string(),
  difficulty: difficultySchema,
  followUp: z.boolean().optional(),
});

export const preparationTimelineStepSchema = z.object({
  day: z.number(),
  label: z.string(),
  phase: z.string(),
  status: z.enum(["complete", "active", "upcoming"]),
});

export const preparationPlanSchema = z.object({
  tasks: z.array(preparationTaskSchema),
  questions: z.array(questionSchema),
  timeline: z.array(preparationTimelineStepSchema),
});

File: Backend/app/schemas/preparation.py
from typing import ClassVar

from app.core.serialization import CamelModel
from app.schemas.common import (
    Difficulty,
    PreparationTaskStatus,
    PreparationTimelineStatus,
    TaskPriority,
)


class PreparationTask(CamelModel):
    id: str
    day: int
    date_label: str
    phase: str
    category: str
    title: str
    description: str
    estimated_minutes: int
    status: PreparationTaskStatus
    priority: TaskPriority


class Question(CamelModel):
    omit_if_none: ClassVar[frozenset[str]] = frozenset({"follow_up"})

    id: str
    text: str
    category: str
    topic: str
    difficulty: Difficulty
    follow_up: bool | None = None


class PreparationTimelineStep(CamelModel):
    day: int
    label: str
    phase: str
    status: PreparationTimelineStatus


class PreparationPlan(CamelModel):
    tasks: list[PreparationTask]
    questions: list[Question]
    timeline: list[PreparationTimelineStep]


class UpdateTaskStatusRequest(CamelModel):
    status: PreparationTaskStatus
